eventmesh/LockManager.js

```diff
@@ -1,105 +1,184 @@
+    opts.lockTTL = opts.lockTTL ? opts.lockTTL : Infinity;
+    _.extend(opts, {
+      'lockTime': currentTime
+    });
+    logger.debug(`Attempting to acquire lock on resource with resourceId: ${resourceId} `);
+    return eventmesh.apiServerClient.getResource(CONST.APISERVER.RESOURCE_GROUPS.LOCK, CONST.APISERVER.RESOURCE_TYPES.DEPLOYMENT_LOCKS, resourceId)
+      .then(resource => {
+        const resourceBody = resource.body;
+        const currentlLockDetails = JSON.parse(resourceBody.spec.options);
+        const currentLockTTL = currentlLockDetails.lockTTL ? currentlLockDetails.lockTTL : Infinity;
+        const currentLockTime = new Date(currentlLockDetails.lockTime);
+        if ((currentTime - currentLockTime) < currentLockTTL) {
+          logger.error(`Resource ${resourceId} was locked for ${currentlLockDetails.lockedResourceDetails.operation} ` +
+            `operation with id ${currentlLockDetails.lockedResourceDetails.resourceId} at ${currentLockTime} `);
+          throw new DeploymentAlreadyLocked(resourceId, {
+            createdAt: currentLockTime,
+            lockForOperation: currentlLockDetails.lockedResourceDetails.operation
           });
         } else {
-          const newLockDetails = {};
-          newLockDetails.count = 1;
-          newLockDetails.operationType = operationType;
-          return client.put(resource + CONST.LOCK_DETAILS_SUFFIX).value(JSON.stringify(newLockDetails));
+          const patchBody = _.assign(resourceBody, {
+            spec: {
+              options: JSON.stringify(opts)
+            }
+          });
+          return eventmesh.apiServerClient.updateResource(CONST.APISERVER.RESOURCE_GROUPS.LOCK, CONST.APISERVER.RESOURCE_TYPES.DEPLOYMENT_LOCKS, resourceId, patchBody);
         }
       })
-      .then(() => {
-        lockDetailsChanged = true;
-        return lock.release();
-      })
-      .catch(e => {
-        if (!lockDetailsChanged) {
-          throw new ETCDLockError(e.message);
-        } else {
-          logger.info('Resource unlock failed. However, now throwing error and letting lock successful as unlock happens automatically after 5 seconds');
+      .tap(() => logger.debug(`Successfully acquired lock on resource with resourceId: ${resourceId}`))
+      .catch(err => {
+        if (err instanceof NotFound) {
+          const spec = {
+            options: JSON.stringify(opts)
+          };
+          const status = {
+            locked: 'true'
+          };
+          const body = {
+            metadata: {
+              name: resourceId
+            },
+            spec: spec,
+            status: status
+          };
+          return eventmesh.apiServerClient.createLock(CONST.APISERVER.RESOURCE_TYPES.DEPLOYMENT_LOCKS, body)
+            .tap(() => logger.debug(`Successfully acquired lock on resource with resourceId: ${resourceId} `));
+        } else if (err instanceof Conflict) {
+          throw new DeploymentAlreadyLocked(resourceId);
```

Comment:
LN 105, creates DeploymentAlreadyLocked by specifying the operation which is running & since when, however here its not being done. Can we put those details here?